File: advent-of-code/src/main/scala/com/willmolloy/adventofcode/_2024/Day8.scala
package com.willmolloy.adventofcode._2024

import com.willmolloy.adventofcode.common.grid.Point
import com.willmolloy.adventofcode.common.{Day, Input}
import com.willmolloy.adventofcode.scala.Adapters.ListAdapter
import com.willmolloy.adventofcode.scala.Adapters.GridAdapter

import scala.collection.mutable
import scala.util.boundary
import scala.util.boundary.break

/**
 * https://adventofcode.com/2024/day/8
 */
object Day8 extends Day(2024, 8) {

  override def part1(input: Input): Long = {
    val grid = input.readCharGrid()
    val nodes = mutable.Set[Point]()

    for (p1 <- grid) {
      for (p2 <- grid) {
        // 2 antenna
        if (p1 != p2 && grid(p1) == grid(p2) && grid(p1) != '.') {
          // math
          val dx = Math.abs(p1.x - p2.x)
          val dy = Math.abs(p1.y - p2.y)

          val x3 = if p1.x < p2.x then p1.x - dx else p1.x + dx
          val y3 = if p1.y < p2.y then p1.y - dy else p1.y + dy
          val p3 = Point(x3, y3)
          if (grid.inbounds(p3)) {
            nodes.add(p3)
          }
        }
      }
    }
    nodes.size
  }

  override def part2(input: Input): Long = {
    val grid = input.readCharGrid()
    val nodes = mutable.Set[Point]()

    for (p1 <- grid) {
      for (p2 <- grid) {
        // 2 antenna
        if (p1 != p2 && grid(p1) == grid(p2) && grid(p1) != '.') {
          // math
          val dx = Math.abs(p1.x - p2.x)
          val dy = Math.abs(p1.y - p2.y)

          var m = 0
          boundary {
            while (true) {
              val x3 = if p1.x < p2.x then p1.x - dx * m else p1.x + dx * m
              val y3 = if p1.y < p2.y then p1.y - dy * m else p1.y + dy * m
              val p3 = Point(x3, y3)
              m += 1
              if (grid.inbounds(p3)) {
                nodes.add(p3)
              } else {
                break()
              }
            }
          }
        }
      }
    }
    nodes.size
  }
}

File: advent-of-code/src/main/kotlin/com/willmolloy/adventofcode/_2025/Day8.kt
package com.willmolloy.adventofcode._2025

import com.willmolloy.adventofcode.common.Day
import com.willmolloy.adventofcode.common.Input
import com.willmolloy.adventofcode.common.datastructures.DisjointSet
import com.willmolloy.adventofcode.common.extensions.product
import com.willmolloy.adventofcode.common.grid.Point3
import java.util.Comparator.comparingLong
import java.util.PriorityQueue

/** https://adventofcode.com/2025/day/8 */
object Day8 : Day(2025, 8) {

  override fun part1(input: Input): Any = solve(input).first

  override fun part2(input: Input): Any = solve(input).second

  private fun solve(input: Input): Pair<Long, Long> {
    // some MST problem
    val points =
      input.readLines().map {
        val split = it.split(",")
        Point3(split[0], split[1], split[2])
      }

    val circuits = DisjointSet(points.size)

    val edges =
      PriorityQueue<Pair<Int, Int>>(comparingLong { points[it.first].dist(points[it.second]) })
    for (src in 0 until points.size) {
      for (dst in src + 1 until points.size) {
        edges.add(src to dst)
      }
    }

    val maxConnections = if (edges.size > 1000) 1000 else 10
    var part1 = 0L
    var part2 = 0L

    for (connections in 0 until edges.size) {
      val edge = edges.remove()

      val src = edge.first
      val dst = edge.second

      circuits.union(src, dst)

      if (connections == maxConnections - 1) {
        val top3 = circuits.sizes().sortedDescending().map { it.toLong() }.take(3)
        part1 = top3.product()
      }

      if (circuits.size(src) == points.size) {
        part2 = (points[src].x * points[dst].x)
        break
      }
    }

    return Pair(part1, part2)
  }
}
